Add more formatting rules and remove risky ones

mrliptontea opened the pull request:
As a sane-defaults project, this config should just focus on formatting code without potentially changing the code behaviour.

I removed all risky rules, including `native_function_invocation`, which as shown #1 doesn't benefit the execution. Users can opt-in to risky rules by tweaking the config.

Closes #1.

Commit:
Add more formatting rules and remove risky ones

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
-        'cast_spaces' => ['space' => 'none'],
-        'combine_consecutive_issets' => true,
-        'combine_consecutive_unsets' => true,
+        'binary_operator_spaces' => true,
```
👎

mrliptontea replied:
Despite its name, rule can be used to align pretty much [all operators](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/53677f035c3c4ebd8d10d95c71f4a15cc6610df4/src/Fixer/Operator/BinaryOperatorSpacesFixer.php#L88-L133), which IMO should be surrounded by a single space.

e.g.

```php
// before
$delta  =$b**2-4 *$a* $c;
// after
$delta = $b ** 2 - 4 * $a * $c;
```

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
-        'combine_consecutive_issets' => true,
-        'combine_consecutive_unsets' => true,
+        'binary_operator_spaces' => true,
+        'cast_spaces' => ['space' => 'single'],
```
👎

mrliptontea replied:
`'single'` is a default and also preferred in [PHP documentation](http://www.php.net/manual/en/language.types.type-juggling.php#language.types.typecasting).

Mezzle replied:
It may be so - however, I tend to go for defaults that protect a user when copying/pasting or refactoring manually.

It'd be quite easy to take just the variable and not the cast when the variable is separated from the cast - hence keeping them together.

Mezzle replied:
still 👎

mrliptontea replied:
While it's not specified in the [extended coding style proposal](https://github.com/php-fig/fig-standards/blob/master/proposed/extended-coding-style-guide.md#6-operators), given that `cast` is a binary operator, it should be followed by a space for consistency.

Also, I don't think copy-paste argument makes sense, because if someone is copying a variable, why would they need to take type casting along? Or, if they did, then casting has a purpose in some wider context, and that would mean more code to the left from cast operator.

Mezzle replied:
Cast isn't a binary operator, it's a unary operator.  it only requires one parameter.

Therefore, it should be treated the same as other unary operators - so we should be doing

```
$i ++;
++ $i;
$x = - 5;
```

if we go by that ruling

mrliptontea replied:
My mistake, the cast operator is unary. Therefore proposal doesn't cover it.

I'm going to change it so we can move on.

mrliptontea replied:
See a51899c7fd8643c735bf5b35dd7b5686949acea7

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
         'list_syntax' => ['syntax' => 'short'],
         'lowercase_cast' => true,
         'magic_constant_casing' => true,
-        'mb_str_functions' => true,
```
👎

mrliptontea replied:
Forcing `mb_` has nothing to do with formatting, is risky, and may introduce unintended behaviour.

Mezzle replied:
Only risky where someone has overridden the functions - which is something that would go against the `sane` in `sane-defaults` :)

Coding standards aren't just about formatting :)

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
-        'modernize_types_casting' => true,
+        'multiline_whitespace_before_semicolons' => true,
         'native_function_casing' => true,
-        'native_function_invocation' => true,
```
👎

mrliptontea replied:
As showed out in #1, it's a pointless rule - all benchmarks show that at best you will save few μs on thousands of invocations, so I see absolutely no reason to force it.

Mezzle replied:
Setting aside the performance stuff here (which is debatable) - the reasoning behind this, again - is to benefit the user when copying and pasting / refactoring..  For the most part - we shouldn't have overriden functions in namesapces - however, some places might have them... Now - what happens when we accidentally have a namespaced function to do with passwords, or similar?  I'd rather make sure that these calls are going where they're intended.

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
         'no_leading_import_slash' => true,
         'no_leading_namespace_whitespace' => true,
-        'no_mixed_echo_print' => ['use' => 'print'],
+        'no_mixed_echo_print' => true,
```
👎

mrliptontea replied:
Never saw anyone who would prefer `print` over `echo`. If you're favouring `print` because of its similarity to `print_r` or `printf` then it's wrong, because the former is a construct and the latter are functions.
Also, `echo` can stream multiple values, like so:

```php
echo 'Saying some words.', PHP_EOL, 'Hello world!';
```

Only advantage of `print` is ability to use it within expressions but I never saw anyone do it.

```php
<?php $condition ? print 'Yes' : null; ?>
```

Mezzle replied:
*shrug* It's personal preference here... I don't know why - but I can probably concede on this.

Mezzle commented on .php_cs:
```diff
@@ -38,46 +38,47 @@ $config->setRules(
+        'no_mixed_echo_print' => true,
         'no_multiline_whitespace_around_double_arrow' => true,
-        'no_multiline_whitespace_before_semicolons' => true,
-        'no_php4_constructor' => true,
```
👎

mrliptontea replied:
It's [deprecated as of PHP 7.0](http://php.net/manual/en/migration70.deprecated.php#migration70.deprecated.php4-constructors). This rule would be only useful if someone is porting a really old project.

Mezzle replied:
Yup - still don't like it... Still want people who've got their head stuck in the dark ages to be aware of it - still dont want E_DEPRECATED notices everywhere ;)

mrliptontea replied:
I'd leave that off. Chances of anyone using this ruleset on that old a project are minuscule, but more importantly it may not produce the correct result:

> Risky when old style constructor being fixed is overridden or overrides parent one.

...in which case it's better that the maintainer of such project sees E_DEPRECATED.

Mezzle replied:
still 👎

mrliptontea replied:
Meh, OK, it doesn't hurt. See 9e99adcc84316001f454931c8be14ef078a88af9.

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
         'no_unneeded_final_method' => true,
         'no_unused_imports' => true,
         'no_useless_else' => true,
-        'no_useless_return' => true,
```
👎

mrliptontea replied:
Personally, I prefer having such return for symmetry.

Mezzle replied:
Not sure what you mean by symmetry here...

I like this rule, because - a good proportion of the time - it's going to make you think about whether the function should be returning something... sane at the end of the method, or if it's void, whether it shouldn't be.

If you've a function where you're returning a value in an if() or similar, but getting to the end of the function and not returning a value, I'd call that a code smell...  If your function returns something, it should return something in all it's code paths... even if it is throwing an exception, or a "Null Object"

mrliptontea replied:
Don't know what I was thinking of. You're right.

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
-        'no_useless_return' => true,
         'no_whitespace_before_comma_in_array' => true,
         'no_whitespace_in_blank_line' => true,
-        'non_printable_character' => ['use_escape_sequences_in_strings' => true],
```
👎

mrliptontea replied:
Generally, I agree but it's a risky rule.

Mezzle replied:
It's risky if you're relying on non-printable characters in your code... This is meant to be sane defaults, - relying on non-printable characters doesn't really fit into sane for me. :)

mrliptontea replied:
I admit I like this rule with `use_escape_sequences_in_strings` option. Will re-add.

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
         'phpdoc_add_missing_param_annotation' => true,
         'phpdoc_indent' => true,
         'phpdoc_no_access' => true,
-        'phpdoc_no_empty_return' => true,
```
👎

mrliptontea replied:
There should always be `@return` in docblock and whether it's null or a void is a valuable information.

Mezzle replied:
I disagree.  IF something isn't returning something, it shouldn't have a return annotation.

mrliptontea replied:
Maybe I'm just used to having those. PHPDoc will assume `@return void` if it's not present and I'm happy with that.

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
         'phpdoc_single_line_var_spacing' => true,
         'phpdoc_trim' => true,
         'phpdoc_types' => true,
-        'pow_to_exponentiation' => true,
```
👎

mrliptontea replied:
I'm generally OK with this transformation except it's a risky rule. `pow()`/`**` isn't used that often anyway, so I'd leave it for users to opt-in.

Mezzle replied:
Again, risky when someone is doing insane things (overriding `pow`)

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
         'phpdoc_trim' => true,
         'phpdoc_types' => true,
-        'pow_to_exponentiation' => true,
-        'protected_to_private' => true,
```
👎

mrliptontea replied:
My bad. c19620b8cc82c3981d2418a3a4f9f282469a34d5

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
         'phpdoc_types' => true,
-        'pow_to_exponentiation' => true,
-        'protected_to_private' => true,
-        'psr4' => true,
```
👎 👎 👎 👎 👎 👎 👎 👎 👎

mrliptontea replied:
My bad. c19620b8cc82c3981d2418a3a4f9f282469a34d5

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
-        'psr4' => true,
-        'random_api_migration' => true,
         'return_type_declaration' => true,
-        'self_accessor' => true,
```
👎

mrliptontea replied:
Generally, I agree but this is a risky rule.

Mezzle commented on .php_cs:
```diff
@@ -86,38 +87,37 @@ $config->setRules(
-        'self_accessor' => true,
         'semicolon_after_instruction' => true,
         'short_scalar_cast' => true,
-        'simplified_null_return' => true,
```
👎

mrliptontea replied:
My bad, restored in c8aed6faccab3d55a6172742330176c1fc98ebc3.

mrliptontea replied:
Actually, NOPE.

Explicit null is better, let's leave empty return only to `void`. Also sometimes you simply can't typehint, because function is meant to return different types, like:

```php
function parse(string $value)
{
    switch (mb_strtolower($value)) {
        case 'true':
            return true;

        case 'false':
            return false;

        case 'null':
            return null; // you wouldn't want to remove null here

        default:
            return $value;
    }
}
```

Commit:
Restore protected_to_private and psr4 rules

Commit:
Restore simplified_null_return rule

Commit:
Allow risky rules by default and restore some

Bring back:
- mb_str_functions
- native_function_invocation
- no_useless_return
- non_printable_character
- pow_to_exponentiation

Commit:
Add blank lines before case: and default: rather than before break;

Commit:
Use lowercase static reference

Commit:
Escape global constant references

Commit:
Disallow binary strings

Commit:
Disallow unsetting properties

Commit:
Align docblocks to the left and remove superfluous blank lines

Commit:
Restore 'no PHP4 constructor' rule

Commit:
Merge branch 'master' into less-invasive-rules

Commit:
Use no space after cast

Commit:
Correct order
